lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,239 @@
+        """Get version of package."""
+        return self._data.get("version", None)
+
+    @property
+    def full_version(self) -> Optional[str]:
+        """Get full version of package."""
+        if self.absent:
+            return None
+
+        full_version = [self.version, f"-{self.release}"]
+        if self.epoch:
+            full_version.insert(0, f"{self.epoch}:")
+
+        return "".join(full_version)
+
+    @property
+    def release(self) -> Optional[str]:
+        """Get release of package."""
+        return self._data.get("release", None)
+
+    @property
+    def repo(self) -> Optional[str]:
+        """Get repository package is from."""
+        return self._data.get("repo", None)
+
+
+def version() -> str:
+    """Get version of `dnf` executable."""
+    return StringIO(_dnf("--version")).readline().strip("\n")
+
+
+def installed() -> bool:
+    """Determine if the `dnf` executable is available on PATH."""
+    return shutil.which("dnf") is not None
+
+
+def update() -> None:
+    """Update all packages on the system."""
+    _dnf("update")
+
+
+def upgrade(*packages: str) -> None:
+    """Upgrade one or more packages.
+
+    Args:
+        *packages (str): Packages to upgrade on system.
+    """
+    if len(packages) == 0:
+        raise ExecutionError("No packages specified.")
+    _dnf("upgrade", *packages)
+
+
+def install(*packages: Union[str, os.PathLike]) -> None:
+    """Install one or more packages.
+
+    Args:
+        *packages (Union[str, os.PathLine]): Packages to install on the system.
+    """
+    if len(packages) == 0:
+        raise TypeError("No packages specified.")
+    _dnf("install", *packages)
+
+
+def remove(*packages: str) -> None:
+    """Remove one or more packages from the system.
+
+    Args:
+        *packages (str): Packages to remove from system.
+    """
+    if len(packages) == 0:
+        raise TypeError("No packages specified.")
+    _dnf("remove", *packages)
+
+
+def purge(*packages: str) -> None:
+    """Purge one or more packages from the system.
+
+    Args:
+        *packages (str): Packages to purge from system.
+    """
+    if len(packages) == 0:
+        raise TypeError("No packages specified.")
+    _dnf("remove", *packages)
+
+
+def fetch(package: str) -> PackageInfo:
+    """Fetch information about a package.
+
+    Args:
+        package (str): Package to get information about.
+
+    Returns:
+        PackageInfo: Information about package.
+    """
+    try:
+        status, info = _dnf("list", "-q", package).split("\n")[:2]  # Only take top two lines.
+        pkg_name, pkg_version, pkg_repo = info.split()
+        name, arch = pkg_name.rsplit(".", 1)
+        epoch, version, release = re.match(r"(?:(.*):)?(.*)-(.*)", pkg_version).groups()
+        if "Installed" in status:
+            state = _PackageState.INSTALLED
+        elif "Available" in status:
+            state = _PackageState.AVAILABLE
+        else:
+            state = _PackageState.UNKNOWN
```

Comment:
> Oddly, in my Fedora VM, when I type `dnf install python2` it works, so it seems to be available. But when I do `dnf list -q python2` it displays "Error: No matching Packages to list" and returns a non-zero exit code. Any idea why, or what category of package `python2` might be? I'm just concerned we're missing a case here.

So I figured out the issue with `python2` on Fedora 36. `dnf install python2` works because dnf resolves the package name to `python2.7`; there is no `python2` package in the package archive, but there is a `python2.7` package that dnf uses instead. Since there is actually no `python2` package but a `python2.7` package, `dnf list -q python2` will say there is no package in existence while `dnf list -q python2.7` will work as expected.

Reply:
Let me know what you think of this new resiliency check here:

https://github.com/canonical/operator-libs-linux/blob/6518bcfbff7e0daa6ec871aba0ee14305f4a1cc8/lib/charms/operator_libs_linux/v0/dnf.py#L169-L190

Rather than just using an _UNKNOWN_ state, I instead raise an error which tells the library to mark the package as absent. If there is faulty output, this means that something is wrong with the package and it should be marked as _ABSENT_. Other possible states are recent and obsolete, but these packages will by default either come up as _AVAILABLE_ or _INSTALLED_. Recent just means that the package was recently modified and obsolete means that there is just a newer version of the package available in the archive.

Reply:
> `dnf install python2` works because dnf resolves the package name to python2.7 ... Since there is actually no python2 package but a python2.7 package, `dnf list -q python2` will say there is no package

Is this going to be a problem with the design of using `dnf list` to get package info? I mean, the fact that `fetch` will return *ABSENT* but `install` can actually install it? Maybe at least worth mentioning in the doc comment.

Reply:
I was using Fedora 37 via `docker run -it fedora` BTW.

Reply:
Sounds good. I replicated the same behavior in my Fedora 36 container via `lxc launch images:fedora/36 fedora-36-test`. I can add a note that the passed package name needs to exactly match the name of the package and cannot be an alias. The same issue is also present using `dnf info python2`:

```
[root@fedora-36-test ~]# dnf info python2
Last metadata expiration check: 2:11:58 ago on Mon Mar  6 20:26:31 2023.
Error: No matching Packages to list
```

Reply:
Added note in https://github.com/canonical/operator-libs-linux/pull/75/commits/d19923ec4e869150f7e257a6c55d8fcffffdaaf5. Reading the `Notes:` section of the docstring will mention that the package name needs to be exact and not an alias.